### PR TITLE
Convert section chars in components to components in more places

### LINKED ISSUE #4888: Item drop Mythicmobs Mobs not combined item paper 1.16.4
### What behaviour is expected:
- Item drop mythicmobs combine item paper

### What behaviour is observed:
- Item drop Mythicmobs not combined item paper 1.16.4 By the way, making it combined is restart server , Enter the server again , load chunk but the name of the item is the same.

### Steps/models to reproduce:

Boss drop in mythicmobs
https://mcpaste.io/866a167ca699e822

- Example Mythicmobs NBT items not combined : https://www.youtube.com/watch?v=OMhOFHD6wyw

### Plugin list:
- AuctionHouse , BetterDrops , ChestShop , Citizens , Clearlag , CMI , Command Panels , EntityTrackerFixer , ItemEdit , Jobs , LockettePro , LuckPerms , MobFarmManager , MythicMobs-4.11.0 , PlaceholderAPI , ProtocolLib , Shopkeepers-2.12.0

### Paper version:
- paper-1.16.4-324 [Lastest Version Paper] Confirm!

### LINKED ISSUE #7599: Items modified while saving
### Expected behavior

When we generate an item with the Adventure API by using :
`Component.text("Bla bla")` for the name and lore, 
We can see by running the command `/data get entity @p SelectedItem` that the item has some data ...

We expect that when we restart the server, items stored in chests has this same data.

### Observed/Actual behavior

Correct data of the item :
image

Incorrect data of the item :
image

This happens when an item generated with Components is stored in a chest and the server is restarted (to force the map to be saved and loaded from disk)

Another effect of this, we can see that we can't stack correct items with incorrect ones.

### Steps/models to reproduce

- Generate an ItemStack with a name and lore generated from "Component.text(...)"
- Observe the data with `/data get entity @p SelectedItem` while holding the ItemStack in your hand
- Place a chest and place the ItemStack inside this chest
- Stop and Start the server
- Get the ItemStack in the chest
- Observe the data with `/data get entity @p SelectedItem`

### Plugin and Datapack List

image

### Paper version

 This server is running Paper version git-Paper-248 (MC: 1.18.2) (Implementing API version 1.18.2-R0.1-SNAPSHOT) (Git: 1ce3555)
You are 1 version(s) behind
Download the new version at: https://papermc.io/downloads
Previous version: git-Paper-181 (MC: 1.18.1)

### Other

This issue may happen with inventories. We are using a plugin for player inventories that saves serialized items in a database. On my server, player's inventory items are saved and read right

### LINKED ISSUE #9399: Escaping edit of sign w/ legacy colors removes legacy colors
### Expected behavior

Signs using legacy colors should be handled the same way as sign with components: When editing the sign and pressing ESC, the sign should not change

### Observed/Actual behavior

When editing a sign with legacy styling and escaping without doing any changes, the text looses all legacy styling

### Steps/models to reproduce

1. Place a sign with legacy colors (I used setblock via console: `setblock -4 64 -6 minecraft:oak_sign[rotation=8,waterlogged=false]{back_text:{color:"black",has_glowing_text:0b,messages:['{"text":""}','{"text":""}','{"text":""}','{"text":""}']},front_text:{color:"black",has_glowing_text:0b,messages:['{"text":"§1test"}','{"text":""}','{"text":""}','{"text":""}']},is_waxed:0b} replace`
2. Right click on the sign
3. Press escape

### Plugin and Datapack List

none

### Paper version

 git-Paper-49 (MC: 1.20.1) (Implementing API version 1.20.1-R0.1-SNAPSHOT) (Git: e8bec64)

### Other

Upstream seems to have fixed this bug for components in https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/66c5ce4c7ba6a32b250c4630eb4c189fc5483542#nms-patches/net/minecraft/world/level/block/entity/TileEntitySign.patch but probably forgot about legacy stuff

### PR DESCRIPTION
**Pull request by Machine-Maker**
Fixes #4888
Fixes #7599
Fixes https://github.com/PaperMC/Paper/issues/9399

Does the same conversion in more places to ensure comparisons are correct. Section chars in top-level components are now converted to json structure. This doesnt touch section chars in children components, but it didnt originally do that either.

### CODE REVIEW

**Comment by Machine-Maker** [comment]
Rebased for 1.18

**Review by electronicboy — APPROVED** [review]

**Comment by Machine-Maker** [comment]
Rebased for 1.18.2

**Comment by Machine-Maker** [comment]
Rebased for 1.19.3

**Comment by Machine-Maker** [comment]
This now also fixes https://github.com/PaperMC/Paper/issues/9399

File: patches/server/0434-Update-itemstack-legacy-name-and-lore.patch [hunk 1]
```diff
@@ -3,12 +3,56 @@ From: William Blake Galbreath <Blake.Galbreath@GMail.com>
...
++    }
++
+     public static Component asAdventure(final net.minecraft.network.chat.Component component) {
+         return component == null ? Component.empty() : GsonComponentSerializer.gson().serializer().fromJson(net.minecraft.network.chat.Component.Serializer.toJsonTree(component), Component.class);
+     }
+@@ -167,9 +172,16 @@ public final class PaperAdventure {
+     }
+ 
+     public static List<String> asJson(final List<? extends Component> adventures) {
++        return PaperAdventure.asJson(adventures, false);
```
Comment (kashike) [inline patches/server/0434-Update-itemstack-legacy-name-and-lore.patch]:
```suggestion
+        return asJson(adventures, false);
```

File: patches/server/0434-Update-itemstack-legacy-name-and-lore.patch [hunk 2]
```diff
@@ -3,12 +3,56 @@ From: William Blake Galbreath <Blake.Galbreath@GMail.com>
...
++    }
++    public static List<String> asJson(final List<? extends Component> adventures, boolean convertLegacy) {
+         final List<String> jsons = new ArrayList<>(adventures.size());
+         for (final Component component : adventures) {
++            if (convertLegacy) {
++                jsons.add(ItemStack.convert(GsonComponentSerializer.gson().serialize(component)));
```
Comment (kashike) [inline patches/server/0434-Update-itemstack-legacy-name-and-lore.patch]: Have the conversion method be in `PaperAdventure`.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by Machine-Maker** [comment]
Closing this as we've already removed any existing logic we had that did this. As more and more move to components, its less and less feasible to try and fix plugin mistakes here. Will be closing the linked issues as well.

**Pull request closed without merging**